### ISSUE TITLE
Implement a caching layer

### ISSUE THREAD

**Issue by Yoric** [matrix-org/mjolnir#374]
One of the problems we encounter with matrix-bot-sdk is that it performs no caching. Consequently, things like:

- calling `getUserId()`;
- calling `getAccountData()`;
- getting state events for a room.

Are:

- slow (they all need to hit the homeserver);
- fallible (in case of network issue, they raise an exception);
- log-heavy (they keep showing up in the logs, over and over again, which makes the logs really hard to parse for a human being).

I propose we create a new class, e.g. `State`, which lets us:

- get the user id synchronously;
- get user data synchronously;
- get (selected) state events for protected rooms synchronously — not *all* state events, as we're only interested in a few.

This class will be initialized by `Mjolnir` and will listen to `/sync` changes to update state events from protected rooms.

**Comment by turt2live** [comment]
I believe the underlying bot-sdk does a lot of this for us now. In general, it's expected that the homeserver is the caching layer for the bot, as the bot does not want to desync from the homeserver on account/room representation.
